src/Domain/LeanCode.DomainModels/Model/TimestampTz.cs

```diff
@@ -0,0 +1,16 @@
+namespace LeanCode.DomainModels.Model;
+
+public sealed record class TimestampTz(DateTime UtcTimestamp, string TimeZoneId) : ValueObject
```

Comment:
I think some helper methods `FromUtc` / `FromLocal` would be nice (with `string` and `TimeZoneInfo` overloads). 
Also possibly time zone should be validated (unless it's a design choice).

Speaking of it, wouldn't it be worth to have a `TimeZoneInfo` property instead of class - it shouldn't complicate much EF and STJ _I hope_

Reply:
> `FromUtc`

Would it be just the same thing as `new(timestamp, "Etc/UTC")`? If so then sure, we can add one.

> `FromLocal`

I see no value in this, shouldn't be too common.

> (with `string` and `TimeZoneInfo` overloads)

What if the `TimeZoneInfo` object we were passed has `HasIanaId == false`? Throw an exception?

> Also possibly time zone should be validated (unless it's a design choice).

This is tricky. How do you verify that passed `TimeZoneId` is a valid IANA ID using only .NET APIs? (and without calling into libicu)

> Speaking of it, wouldn't it be worth to have a `TimeZoneInfo` property instead of class - it shouldn't complicate much EF and STJ _I hope_

I think it would complicate EF, because afaik it's not possible to store `TimeZoneInfo` directly in a column without writing a converter and with one, rewriting expressions could become much harder than it is right now. Also, how is STJ relevant here?

Reply:
> > `FromLocal`
> 
> I see no value in this, shouldn't be too common.

I mean `FromLocal(DateTime/DateTimeOffset, TimeZoneInfo)`, not the local machine time- this seems to be a common use case - do you expect clients to always send utc + tz pair? 

>> (with string and TimeZoneInfo overloads)
>
> What if the TimeZoneInfo object we were passed has HasIanaId == false? Throw an exception?

What if I pass `XD` string as timezone id? How would postgres `AT TIME ZONE` react to that? I mean, we'd need to validate this at some point.

>>  Also possibly time zone should be validated (unless it's a design choice).
>
> This is tricky. How do you verify that passed TimeZoneId is a valid IANA ID using only .NET APIs? (and without calling into libicu)

What is wrong with .NET APIs? Is it not up to date / enough for our use cases? I think usually it's the backend that dictates available time zones anyway.

> > Speaking of it, wouldn't it be worth to have a `TimeZoneInfo` property instead of class - it shouldn't complicate much EF and STJ _I hope_
> 
> I think it would complicate EF, because afaik it's not possible to store `TimeZoneInfo` directly in a column without writing a converter and with one, rewriting expressions could become much harder than it is right now. Also, how is STJ relevant here?

Okay, yeah this would call for custom translation? Speaking of `STJ` - this VO may still be serialized in events, json columns, etc.

Reply:
> I mean `FromLocal(DateTime/DateTimeOffset, TimeZoneInfo)`, not the local machine time- this seems to be a common use case - do you expect clients to always send utc + tz pair?

No, I'd expect the client to always send timezone ID and either some kind of `DateTimeOffset` (because contracts forbid `DateTime`), in which case using the constructor is about as complicated as calling `.UtcDateTime` on that `DateTimeOffset`. If you believe that deserves a helper method, sure, I can add one.

> What if I pass `XD` string as timezone id? How would postgres `AT TIME ZONE` react to that? I mean, we'd need to validate this at some point.

We could, but it will never be foolproof. Let me know when you figure out how to extract IANA ID (and not just any ID that could be a Windows ID) from any arbitrary `TimeZoneInfo` object and maybe we will give this a try.

> What is wrong with .NET APIs? Is it not up to date / enough for our use cases? I think usually it's the backend that dictates available time zones anyway.

I recall there being some project that attempted to validate timezone IDs only to fail horribly when passed non-canonical/linked IDs or Windows IDs.

> Okay, yeah this would call for custom translation? Speaking of `STJ` - this VO may still be serialized in events, json columns, etc.

I'm just not sure if EF won't scream at us for attempting to construct an method expression that expects a parameter of type `string` but is given a parameter of type `TimeZoneInfo` and backend isn't aware that `TimeZoneInfo` is actually a lie.

About STJ, good point, less about JSON columns but certainly about events. At worst we will add a new converter to `LeanCode.Serialization` but I'll try to check first it it would be needed at all.

Reply:
OK, having the rest resolved let's focus on just `TimeZoneInfo` validation applicability.

I believe accepting plain string without any validation is just escaping from the problem. _At some point_ you'll need to convert something, either client/backend/db and you'll encounter undefined behaviour. 

I agree that is might be easy to shoot oneself in the feet, probably we'd need to extract validation to some static class that can be overridden in the project if necessary. 

 > We could, but it will never be foolproof. Let me know when you figure out how to extract IANA ID (and not just any ID that could be a Windows ID) from any arbitrary TimeZoneInfo object and maybe we will give this a try.

Isn't `TimeZoneInfo.TryConvertWindowsIdToIanaId` + `TimeZoneInfo.HasIanaId` enough / enough to filter out all the zones that we don't need? Also, in some of the projects I took an approach where the backend returns supported timezones to clients (clients send local time and tz id and can even be unaware of the offsets write time).

I'm leaning toward some `ITimeZoneProvider` interface with sensible defaults, that will list all supported timezones + let you validate id.

Reply:
I implemented validation of time zone IDs but I'd say I'm against making it configurable. I don't think there are any reasonable alternatives other than what .NET already provides via libicu.

Let's hope nobody will try to run this code in minimal Alpine container.

Reply:
I mean, this could be an escape hatch if you really need to override this but I can do without.